### PR TITLE
Add validation for hostnames in HTTPRoute

### PR DESCRIPTION
**Pull request by nak3**
This patch adds validaiton for HTTPRoute.

The requirement is same with
`ingresses.networking.k8s.io.spec.rules.host` (please refer to it by
`$ kubectl explain ingresses.networking.k8s.io.spec.rules.host`) so
this patch just adds the same code from [here](https://github.com/kubernetes/kubernetes/blob/3b5d809eedcc60e1e8fe63dace1f453d6a27b949/pkg/apis/networking/validation/validation.go#L346-L360).

Fix https://github.com/projectcontour/contour/issues/3353

Signed-off-by: Kenjiro Nakayama <nakayamakenjiro@gmail.com>

/cc @stevesloka @sunjayBhatia @danehans

### CODE REVIEW

**Review by stevesloka — APPROVED** [review]

**Review by sunjayBhatia — COMMENTED** [review]
Should we leave a TODO to set some status on errors?

If we are going with the whatever plans to do Gateway validation in the operator will we need to duplicate this logic there?

Also could maybe add a test or two for some more edge cases, like wildcard domains with the `*` in different places, a single DNS label for the hostname (like we might have for a k8s Service), etc.

**Comment by danehans** [comment]
Since the same validation needs to be performed for Gateway, we should do either:
1. Create a contour/pkg/validation pkg that includes the logic in this PR.
2. If the operator will responsible for reconciling a Gateway (which includes the same hostname field), this logic should be in repo that contains shared packages.

xref the same validation implemented by the operator for Gateway hostnames: https://github.com/projectcontour/contour-operator/pull/226

**Comment by danehans** [comment]
/lgtm

I'll submit a follow-on PR to move this logic into a validation pkg.

**Review by danehans — APPROVED** [review]

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -114,6 +118,22 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1alpha1.HTTPRo
...
+				continue
+			}
+			if strings.Contains(hostname, "*") {
+				if errs := validation.IsWildcardDNS1123Subdomain(hostname); errs != nil {
+					p.Errorf("invalid hostname %q: %v", hostname, errs)
+					continue
+				}
+			} else {
+				if errs := validation.IsDNS1123Subdomain(hostname); errs != nil {
+					p.Errorf("invalid listener hostname %q", hostname, errs)
```
Comment (jpeach) [inline internal/dag/gatewayapi_processor.go:133]:
```suggestion
p.Errorf("invalid listener hostname %q: %v", hostname, errs)
```
